### PR TITLE
special-case optional any properties

### LINKED ISSUE #287: Optional fields with any type not handled correctly
Currently tscikle translates 

```ts
interface I {
  a?: any
}
```
to 

```javascript
/** @record */
function I(){};

/** @type {?} */
I.prototype.a;
```

That is a problem because the following code passes TS, but errors in closure:
``` let x: I = {}``` translated to:
```
/** @type {I} */
var x = {};
```

For closure, tsickle should output:

```
/** @type {?|undefined} */
I.prototype.a;
```

Admittedly, one might expect '? | T' to be always the same as '?', so this might be a closure bug? 

//cc @dpurp

### PR DESCRIPTION
**Pull request by evmar**
When a property is optional, e.g.
  foo?: string;
Then the type of the property is string|undefined and the above
translation handles it correctly.
But in the special case of any, which includes undefined:
  foo?: any;
The type is just "any" (because that covers undefined as well) and
our translation of the type is just "?".  But Closure wants it to
instead be ?|undefined.

Fixes #287.

### CODE REVIEW

File: src/tsickle.ts [hunk 1]
```diff
@@ -800,7 +800,7 @@ class Annotator extends ClosureRewriter {
 
     const memberNamespace = [name, 'prototype'];
     for (let elem of iface.members) {
-      this.visitProperty(memberNamespace, elem);
+      this.visitProperty(memberNamespace, elem, elem.questionToken != null);
```
Comment (mprobst) [inline src/tsickle.ts]: nit: add a comment `/**optional=*/` or use named parameters?

File: test_files/basic.untyped/basic.untyped.tsickle.ts [hunk 2]
```diff
@@ -17,9 +17,9 @@ constructor(private ctorArg: string) {
 }
 
 function Foo_tsickle_Closure_declarations() {
-/** @type {?} */
+/** @type {?|undefined} */
```
Comment (mprobst) [inline test_files/basic.untyped/basic.untyped.tsickle.ts]: is this an intentional/good change? the original field was just `string`, this seems to slip in due to being untyped?

File: src/tsickle.ts [hunk 3]
```diff
@@ -867,15 +867,31 @@ class Annotator extends ClosureRewriter {
...
+    // Interesting special case:
+    // When a property is optional, e.g.
+    //   foo?: string;
+    // Then the type of the property is string|undefined and the above
+    // translation handles it correctly.
+    // But in the special case of any, which includes undefined:
+    //   foo?: any;
+    // The type is just "any" (because that covers undefined as well) and
+    // our translation of the type is just "?".  But Closure wants it to instead
+    // be ?|undefined.
```
Comment (mprobst) [inline src/tsickle.ts]: Poking at Closure Compiler suggests that `{?}` includes `undefined`. Is this only needed in a particular special case, like leaving out the properties of an interface or so?

https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F%252F%2520ADD%2520YOUR%2520CODE%2520HERE%250A%252F**%2520%2540type%2520%257Bundefined%257D%2520*%252F%250Avar%2520undef%253B%250A%252F**%2520%2540type%2520%257B%253F%257D%2520*%252F%250Avar%2520x%2520%253D%2520undef%253B%250Ax%2520%253D%25201%253B%250Aundef%2520%253D%25202%253B

Comment (evmar) [inline src/tsickle.ts]: Yep, exactly -- see the review description :P

Comment (mprobst) [inline src/tsickle.ts]: > But Closure wants it to instead be ?|undefined.

[citation required]. Or explanation or so. See the link I sent above, if you type a variable as `?` in Closure, you can assign something typed `undefined` into it.

Comment (evmar) [inline src/tsickle.ts]: I think you overlooked: "When a *property is optional*".


https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%2520%2540record%2520*%252F%250Afunction%2520Foo()%2520%257B%257D%250A%250A%252F**%2520%2540type%2520%257B%253F%257D%2520*%252F%250AFoo.prototype.bar%253B%250A%250A%252F**%2520%2540type%2520%257B%253F%257Cundefined%257D%2520*%252F%250AFoo.prototype.baz%253B%250A%250A%252F**%2520%2540param%2520%257B!Foo%257D%2520foo%2520*%252F%250Afunction%2520useFoo(foo)%2520%257B%2520console.log(foo)%253B%2520%257D%250A%250A%252F%252F%2520illegal%253A%250AuseFoo(%257B%257D)%253B%250AuseFoo(%257Bbaz%253A%25203%257D)%253B%250A%250A%252F%252F%2520legal%253A%250AuseFoo(%257Bbar%253A%25203%257D)%253B

Comment (mprobst) [inline src/tsickle.ts]: Ah that makes sense then.

**Comment by evmar** [comment]
PTAL

File: src/tsickle.ts (hunk 3 above)
Comment (mprobst) [inline src/tsickle.ts]: Coming back to this code review, I *again* stumbled over this comment. Can you change `But Closure wants it to instead be ?|undefined.` to `But for properties (unlike e.g. variables) Closure wants it to instead be ?|undefined.`? I don't blame your comment writing, it's just that the behaviour is so strange.

**Comment by mprobst** [comment]
Friendly ping. A user recently ran into this.

**Comment by evmar** [comment]
PTAL.  Also I'm gonna need to sync current tsickle into google3 before I land this so I guess it's maybe two days out -- one day to run presubmits for current tsickle, then another to run presubmits for tsickle with this change.

**Pull request merged**